### PR TITLE
Implement campfires

### PR DESCRIPTION
**Pull request by xNatsuri**
this pr implements campfires 

 built off #596 just updated

### CODE REVIEW

File: server/block/campfire.go [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+	}
+}
+
+// allCampfires ...
+func allCampfires() (campfires []world.Block) {
+	for _, d := range cube.Directions() {
+		for _, e := range []bool{true, false} {
+			campfires = append(campfires, Campfire{Facing: d, Extinguished: e, Type: NormalFire()})
+			campfires = append(campfires, Campfire{Facing: d, Extinguished: e, Type: SoulFire()})
+		}
```
Comment (DaPigGuy) [inline server/block/campfire.go:263]:
```suggestion
		for _, f := range FireTypes() {
			campfires = append(campfires, Campfire{Facing: d, Type: f, Extinguished: true})
			campfires = append(campfires, Campfire{Facing: d, Type: f})
		}
```

File: server/item/recipe/recipe.go [hunk 2]
```diff
@@ -59,6 +59,20 @@ func NewSmithingTrim(base, addition, template Item, block string) SmithingTrim {
 	}}
 }
 
+// FurnaceTransform represents a recipe only craftable in a furnace.
+type FurnaceTransform struct {
```
Comment (DaPigGuy) [inline server/item/recipe/recipe.go:63]: How about calling it Furnace?

File: server/item/recipe/recipe.go [hunk 3]
```diff
@@ -59,6 +59,20 @@ func NewSmithingTrim(base, addition, template Item, block string) SmithingTrim {
...
+	recipe
+}
+
+// NewFurnaceRecipe creates a new furnace recipe and returns it.
+func NewFurnaceRecipe(input Item, output item.Stack, block string) FurnaceTransform {
```
Comment (DaPigGuy) [inline server/item/recipe/recipe.go:68]: Make sure this matches the name of the struct, e.g. `Shapeless`/`NewShapeless`

File: server/item/recipe/item.go [hunk 4]
```diff
@@ -60,6 +105,45 @@ func (d inputItems) Items() ([]Item, bool) {
 	return s, true
 }
 
+// outputItem ...
+type outputItem struct {
```
Comment (DaPigGuy) [inline server/item/recipe/item.go:109]: You can reuse this type for the `outputItems` type

File: server/item/recipe/item.go [hunk 5]
```diff
@@ -16,6 +16,51 @@ type Item interface {
 	Empty() bool
 }
 
+// inputItems ...
```
Comment (DaPigGuy) [inline server/item/recipe/item.go:19]:
```suggestion
// inputItem ...
```

Reuse this type for `inputItems`

File: server/item/recipe/item.go [hunk 6]
```diff
@@ -60,6 +105,45 @@ func (d inputItems) Items() ([]Item, bool) {
...
+		Name       string                 `nbt:"name"`
+		Properties map[string]interface{} `nbt:"states"`
+		Version    int32                  `nbt:"version"`
+	} `nbt:"block"`
+	// NBTData contains extra NBTData which may modify the item in other, more discreet ways.
+	NBTData map[string]interface{} `nbt:"data"`
+}
+
+// Stack ...
+func (o outputItem) Stack() (item.Stack, bool) {
```
Comment (DaPigGuy) [inline server/item/recipe/item.go:127]: Reuse this for `outputItems.Stack`

e.g

```go
func (d outputItems) Stacks() ([]item.Stack, bool) {
	s := make([]item.Stack, 0, len(d))
	for _, o := range d {
		stack, ok := o.Stack()
		if !ok {
			return nil, false
		}
		s = append(s, stack)
	}
	return s, true
```

File: server/item/recipe/item.go [hunk 7]
```diff
@@ -16,6 +16,51 @@ type Item interface {
...
+		Name       string                 `nbt:"name"`
+		Properties map[string]interface{} `nbt:"states"`
+		Version    int32                  `nbt:"version"`
+	} `nbt:"block"`
+	// Tag is included if the input item is defined by a tag instead of a specific item.
+	Tag string `nbt:"tag"`
+}
+
+// Item ..
+func (i inputItem) Item() (Item, bool) {
```
Comment (DaPigGuy) [inline server/item/recipe/item.go:38]: Reuse for `inputItems.Items`

**Review by DaPigGuy — CHANGES_REQUESTED** [review]

File: server/item/recipe/item.go [hunk 8]
```diff
@@ -16,8 +16,8 @@ type Item interface {
 	Empty() bool
 }
 
-// inputItems is a type representing a list of input items, with helper functions to convert them to
-type inputItems []struct {
+// inputItem is a type representing an input item, with helper function to convert it to
```
Comment (DaPigGuy) [inline server/item/recipe/item.go:19]: convert it to ...?

File: server/item/recipe/item.go [hunk 9]
```diff
@@ -34,34 +34,51 @@ type inputItems []struct {
 	Tag string `nbt:"tag"`
 }
 
+// Item converts input item to recipe item.
```
Comment (DaPigGuy) [inline server/item/recipe/item.go:37]:
```suggestion
// Item converts an input item to a recipe item.
```

File: server/item/recipe/item.go [hunk 10]
```diff
@@ -78,23 +95,39 @@ type outputItems []struct {
 	NBTData map[string]interface{} `nbt:"data"`
 }
 
+// Stack converts output item to item stack.
```
Comment (DaPigGuy) [inline server/item/recipe/item.go:98]:
```suggestion
// Stack converts an output item to an item stack.
```

**Review by DaPigGuy — CHANGES_REQUESTED** [review]

File: server/item/recipe/item.go [hunk 11]
```diff
@@ -34,34 +34,51 @@ type inputItems []struct {
 	Tag string `nbt:"tag"`
 }
 
+// Item converts an input item to a recipe item.
+func (i inputItem) Item() (Item, bool) {
+	var d Item
```
Comment (DaPigGuy) [inline server/item/recipe/item.go:39]: This variable is not needed, just return the Item directly

File: server/block/campfire.go [hunk 12]
```diff
@@ -0,0 +1,267 @@
...
+		if flammable.OnFireDuration() > 0 && c.Extinguished {
+			c.Extinguished = false
+			w.PlaySound(pos.Vec3(), sound.Ignite{})
+			w.SetBlock(pos, c, nil)
+		}
+		if !c.Extinguished && !w.RainingAt(pos) {
+			if flammable.OnFireDuration() < time.Second*8 {
+				flammable.SetOnFire(8 * time.Second)
+			}
+		}
```
Comment (DaPigGuy) [inline server/block/campfire.go:204]: I don't see this behavior in vanilla or on the wiki

See https://minecraft.wiki/w/Campfire#Damage and https://minecraft.wiki/w/Soul_Campfire#Damage

**Review by DaPigGuy — CHANGES_REQUESTED** [review]

**Review by DaPigGuy — CHANGES_REQUESTED** [review]

File: server/block/campfire.go [hunk 13]
```diff
@@ -112,6 +112,14 @@ func (c Campfire) Activate(pos cube.Pos, _ cube.Face, w *world.World, u item.Use
 	if held.Empty() {
 		return false
 	}
+
+	if _, ok := held.Item().(item.Shovel); ok {
```
Comment (DaPigGuy) [inline server/block/campfire.go:116]: This should take durability from the shovel. You can implement this interface to get that automatically: https://github.com/df-mc/dragonfly/blob/master/server/item/shovel.go#L41

Comment (DaPigGuy) [reply server/block/campfire.go:116]: Also check first that the campfire isn't already extinguished

File: server/block/campfire.go [hunk 14]
```diff
@@ -0,0 +1,281 @@
...
+			w.PlaySound(pos.Vec3(), sound.Ignite{})
+			w.SetBlock(pos, c, nil)
+		}
+		if !c.Extinguished {
+			if l, ok := e.(livingEntity); ok && !l.AttackImmune() {
+				for _, t := range FireTypes() {
+					if c.Type == t {
+						l.Hurt(t.Damage(), FireDamageSource{})
+					}
+				}
```
Comment (DaPigGuy) [inline server/block/campfire.go:216]: ?...

```suggestion
				l.Hurt(c.Type.Damage(), FireDamageSource{})
```

File: server/block/campfire.go [hunk 15]
```diff
@@ -0,0 +1,281 @@
...
+	if updated {
+		w.SetBlock(pos, c, nil)
+	}
+}
+
+// NeighbourUpdateTick ...
+func (c Campfire) NeighbourUpdateTick(pos, _ cube.Pos, w *world.World) {
+	_, ok := w.Liquid(pos)
+	_, okTwo := w.Liquid(pos.Side(cube.FaceUp))
+	if ok || okTwo && !c.Extinguished {
```
Comment (DaPigGuy) [inline server/block/campfire.go:195]: Only water extinguishes the campfire

**Review by DaPigGuy — CHANGES_REQUESTED** [review]

**Review by TwistedAsylumMC — APPROVED** [review]

File: server/block/campfire.go [hunk 16]
```diff
@@ -0,0 +1,280 @@
...
+	if updated {
+		w.SetBlock(pos, c, nil)
+	}
+}
+
+// NeighbourUpdateTick ...
+func (c Campfire) NeighbourUpdateTick(pos, _ cube.Pos, w *world.World) {
+	_, ok := w.Liquid(pos)
+	liquid, okTwo := w.Liquid(pos.Side(cube.FaceUp))
+	if ok || okTwo && liquid.LiquidType() == "water" && !c.Extinguished {
```
Comment (DaPigGuy) [inline server/block/campfire.go:198]: Use parenthesis to make things clearer

**Review by DaPigGuy — CHANGES_REQUESTED** [review]

File: server/item/recipe/item.go [hunk 17]
```diff
@@ -16,8 +16,8 @@ type Item interface {
 	Empty() bool
 }
 
-// inputItems is a type representing a list of input items, with helper functions to convert them to
-type inputItems []struct {
+// inputItem is a type representing an input item, with helper function to convert it to an Item.
```
Comment (DaPigGuy) [inline server/item/recipe/item.go:19]:
```suggestion
// inputItem is a type representing an input item, with a helper function to convert it to an Item.
```

File: server/block/campfire.go [hunk 18]
```diff
@@ -0,0 +1,279 @@
...
+	"math/rand"
+	"strconv"
+	"time"
+)
+
+// Campfire is a block that can be used to cook food, pacify bees, act as a spread-proof light source, smoke signal or
+// damaging trap block.
+type Campfire struct {
+	bass
+	sourceWaterDisplacer
```
Comment (DaPigGuy) [inline server/block/campfire.go:21]: transparent

**Review by DaPigGuy — CHANGES_REQUESTED** [review]

File: server/block/campfire.go
Comment (DaPigGuy) [inline server/block/campfire.go:1]: Not documented on the wiki, but for Bedrock Edition, the cooking time is reset on extinguish 
* _and Java Edition loses cooking time at a 2x rate per extinguished tick but the wiki isn't that detailed_ :sweat_smile:

File: server/entity/splashable.go [hunk 19]
```diff
@@ -10,6 +10,20 @@ import (
...
+	world.Block
+	// Splash is called when a type that implements splashable splashes onto a block.
+	Splash(w *world.World, pos cube.Pos, p potion.Potion)
+}
+
+// SplashableEntity is an entity that can be splashed with a splash bottle.
+type SplashableEntity interface {
+	world.Entity
+	// Splash is called when a type that implements splashable splashes onto an entity.
+	Splash(w *world.World, pos mgl64.Vec3, p potion.Potion)
```
Comment (DaPigGuy) [inline server/entity/splashable.go:24]: The `potion.Potion`  passed will always be `potion.Water()`. I'm not entirely sure if there is much of a need for this outside of water splash potions though, so maybe just get rid of the third arg?

cc @TwistedAsylumMC

File: server/entity/splashable.go [hunk 20]
```diff
@@ -66,10 +80,25 @@ func potionSplash(durMul float64, pot potion.Potion, linger bool) func(e *Ent, r
...
+
+				resultPos := result.BlockPosition()
+				if b, ok := w.Block(resultPos).(SplashableBlock); ok {
+					b.Splash(w, resultPos, pot)
 				}
 			case trace.EntityResult:
 				// TODO: Damage endermen, blazes, striders and snow golems when implemented and rehydrate axolotls.
 			}
+
+			for _, otherE := range w.EntitiesWithin(box.Grow(3), ignores) {
```
Comment (DaPigGuy) [inline server/entity/splashable.go:97]: Where is this bounding box size coming from? Shouldn't this match the above used for splash potions with effects?

File: server/block/campfire.go [hunk 21]
```diff
@@ -0,0 +1,280 @@
...
+
+// UseOnBlock ...
+func (c Campfire) UseOnBlock(pos cube.Pos, face cube.Face, _ mgl64.Vec3, w *world.World, user item.User, ctx *item.UseContext) (used bool) {
+	if _, ok := w.Block(pos).(Campfire); ok && face == cube.FaceUp {
+		return false
+	}
+	pos, _, used = firstReplaceable(w, pos, face, c)
+	if !used {
+		return
+	}
```
Comment (DaPigGuy) [inline server/block/campfire.go:156]: Looks like Bedrock Edition just generally does not let you place a campfire on top of another campfire, regardless of which face was tapped

```suggestion
	pos, _, used = firstReplaceable(w, pos, face, c)
	if !used {
		return
	}
	if _, ok := w.Block(pos.Side(cube.FaceDown)).(Campfire); ok {
		return false
	}
```

**Review by DaPigGuy — CHANGES_REQUESTED** [review]

File: server/entity/splashable.go (hunk 19 above)
Comment (xNatsuri) [reply server/entity/splashable.go:24]: I keep this the same until seb reviews it but I don't see a use for it either.

Comment (TwistedAsylumMC) [reply server/entity/splashable.go:24]: Happy for the 3rd arg to be removed here

File: server/block/campfire.go
Comment (DaPigGuy) [reply server/block/campfire.go:1]: You missed the extinguish on splash. Also, make an extinguish function to reduce code duplication

File: server/entity/splashable.go (hunk 19 above)
Comment (DaPigGuy) [reply server/entity/splashable.go]: Do the same for the SplashableBlock interface

File: server/entity/splashable.go [hunk 22]
```diff
@@ -10,6 +10,20 @@ import (
 	"time"
 )
 
+// SplashableBlock is a block that can be splashed with a splash bottle.
+type SplashableBlock interface {
+	world.Block
+	// Splash is called when a type that implements splashable splashes onto a block.
```
Comment (DaPigGuy) [inline server/entity/splashable.go:16]:
```suggestion
	// Splash is called when a water bottle splashes onto a block.
```

and same for SplashableEntity

**Review by DaPigGuy — APPROVED** [review]

**Pull request merged**